File: Minimum Domino rotations for equal row.cpp
//Time Complexity-O(n)-->'n' is size of input vectors
//Space Complexity-O(1)
//Did the code execute on Leetcode? Yes

class Solution {
public:
    
    int calculate(vector<int>& A, vector<int>& B,int x)
    {
        int count1=0;
        int count2=0;
        for(int i=0;i<A.size();i++)
        {
            if(A[i]!=x && B[i]!=x)
            {
                return -1;
            }
            else if(A[i]!=x)
            {
                count1++;
            }
            else if(B[i]!=x)
            {
                count2++;
            }
        }
        return min(count1,count2);
    }
    
    int minDominoRotations(vector<int>& A, vector<int>& B) {
        if(A.size()==0 && B.size()==0)
        {
            return 0;
        }
        int res=calculate(A,B,A[0]);
        if(res!=-1)
        {
            return res;
        }
        else
        {
            return calculate(A,B,B[0]);
        }
    }
};

File: MinimumDominoRotationsForEqualRow.py
# Time Complexity : O(N)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Your code here along with comments explaining your approach
# Using Greedy Approach


class Solution:
    def minDominoRotations(self, A: List[int], B: List[int]) -> int:
        if not A or not B:
            return None
        maxfreq = -inf
        for i in range(1, 7 + 1):
            count = 0
            for j in range(len(A)):
                if A[j] == i:
                    count += 1
                if B[j] == i:
                    count += 1
            if count >= len(A):
                maxfreq = i
                break

        aRotations = 0
        bRotations = 0
        for i in range(len(A)):
            if A[i] != maxfreq and B[i] != maxfreq:
                return -1
            if A[i] != maxfreq:
                aRotations += 1
            if B[i] != maxfreq:
                bRotations += 1
        return min(aRotations, bRotations)
